src/thunderbots/software/util/parameter/parameter.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <ros/ros.h>
+#include <memory>
+#include <string>
+#include <vector>
+
+/**
+ * This class defines a dynamic parameter, meaning the parameter
+ * value can be changed during runtime. Although the class is templated, it is only meant
+ * to support the types also supported by the ROS Parameter Server.
+ *
+ * See http://wiki.ros.org/Parameter%20Server for the list of types
+ *
+ * In our codebase, we support bool, int32_t, double, string, and lists (aka vectors)
+ * of the previous 4 types
+ */
+template <class T>
+class Parameter
+{
+   public:
+    /**
+     * Constructs a new Parameter
+     *
+     * @param ros_parameter_path The global path in the ROS parameter server where this
+     * parameter is stored
+     * @param default_value The default value for this parameter
+     */
+    explicit Parameter<T>(const std::string& ros_parameter_path, T default_value)
+    {
+        this->ros_parameter_path = ros_parameter_path;
+        value_                   = default_value;
+
+        Parameter<T>::registerParameter(std::shared_ptr<Parameter<T>>(this));
```

Comment:
I'm semi-ok with this here, just because all the functions are short, but generally we should separate out declaration from definition (ie. put the definition in another file) even for templated functions.

Reply:
Ah I guess that's where `.tpp` files usually come into play. I'm going to leave it together in one file for now.